src/gutter-container.coffee

```diff
@@ -0,0 +1,131 @@
+{Model} = require 'theorist'
+{CompositeDisposable, Emitter} = require 'event-kit'
+{Gutter} = require './gutter'
+
+# This class encapsulates the logic for adding and modifying the gutters on a
+# TextEditor.
+
+module.exports =
+class GutterContainer extends Model
+  # @param textEditor (required) The TextEditor that this GutterContainer is
+  #   associated with.
+  constructor: (textEditor) ->
+    @textEditor = textEditor
+    @leftGutters = []
+    @rightGutters = []
+    @defaultGutter = null
+    @emitter = new Emitter
+
+  destroy: ->
+    @leftGutters = null
+    @rightGutters = null
+    @defaultGutter = null
+    @emitter.dispose()
+    super()
+
+  # Adds a gutter to the left side of the line numbers. Returns a Gutter instance.
+  # @param gutterName (required) a unique string that identifies this gutter in
+  #   the current TextEditor. If the name is already in use, an error is thrown.
+  # @param options (optional) An object with the following fields:
+  #   priority: Determines stacking order between gutters. Lower priority items
+  #     are forced closer to the edges of the window. (default: 100)
+  #   visibleInitially: boolean specifying whether the gutter is visible
+  #     after being created. Defaults to true.
+  addLeftGutter: (gutterName, options) ->
+    options = options ? {}
+    options.location = 'left'
+    @_addSideGutter gutterName, options, @leftGutters
+
+  # Adds a gutter to the right side of the line numbers. Returns a Gutter instance.
+  # @param are the same as for 'addLeftGutter'
+  addRightGutter: (gutterName, options) ->
+    options = options ? {}
+    options.location = 'right'
+    @_addSideGutter gutterName, options, @rightGutters
+
+  _createGutter: (gutterName, options) ->
+    existingGutter = @_getExistingGutterWithName gutterName
+    if existingGutter
+      throw new Error 'Tried to create a gutter with a name that is already in use.'
+    newGutter = new Gutter gutterName, this, options
+    @subscribe newGutter.onHide (gutter) =>
+      @emitter.emit 'did-hide-gutter', gutter
+    @subscribe newGutter.onShow (gutter) =>
+      @emitter.emit 'did-show-gutter', gutter
+    newGutter
+
+  _addSideGutter: (gutterName, options, gutterArray) ->
+    newGutter = @_createGutter gutterName, options
+    @_insertSortedByPriority newGutter, gutterArray
+    newGutter
+
+  getLeftGutters: ->
+    @leftGutters
+
+  getRightGutters: ->
+    @rightGutters
+
+  getDefaultGutter: ->
+    @defaultGutter
+
+  # This sets the gutter that sits between the left gutters and right gutters.
+  # @param options (optional) An object with the following fields:
+  #   visibleInitially: boolean specifying whether the gutter is visible
+  #     after being created. Defaults to true.
+  setDefaultGutter: (gutterName, options) ->
+    options = options ? {}
+    options.location = 'center'
+    @defaultGutter = @_createGutter gutterName, options
+
+  _getExistingGutterWithName: (name) ->
```

Comment:
Do gutters actually need names? I can see it being useful for documentation purposes but I wonder if gutter decorations should be associated back to their gutters with some kind of auto-incrementing id rather than a name that has to be unique.